### PR TITLE
The Monterey Merge

### PR DESCRIPTION
**Pull request by millenomi**
This PR brings the FOSS portions of Core Foundation on par with with macOS Monterey, iOS 15, watchOS 8 and tvOS 15. This includes the changes shipped during the previous releases.

Most of the changes include bug fixes and performance improvements, but some are useful to highlight:

 - This patch adds the Core Foundation code that underlies Foundation's ListFormatter and RelativeDateTimeFormatter to the FOSS subset.

 - Several Emoji fixes are included that update CFString (and thus NSString and the Foundation methods on Swift.String) support for correctly decoding and iterating over emoji introduced during the Big Sur and Monterey timeframes. The corresponding character maps in CharacterSet have been updated.

 - Certain parsing functions, most notably those that implement property list parsing, now have a maximum recursion depth to prevent resource abuse from malicious input.

 - Some architectural changes have been introduced that are used on Darwin to support Pointer Authentication. For more information on the topic, see [Preparing Your App to Work with Pointer Authentication](https://developer.apple.com/documentation/security/preparing_your_app_to_work_with_pointer_authentication). Note that the FOSS subset does not ship on architectures that require this support, but the architectural changes have been merged for simplicity and synchronization between Darwin and FOSS builds.

 - Parameter checking has been improved for several CF functions.

 - Some implementations have been streamlined or consolidated, especially for CFPlugIn and CFBundle. CFBundle also adds support for "Wrapper"-style bundles (for example, the kind of bundle produced when installing iOS applications on Macs with Apple silicon).

### CODE REVIEW

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
cc @parkera

**Comment by millenomi** [comment]
cc @compnerd — This has a few bits that I know we need Windows expertise on; can we kick testing and see if there's anything that breaks?

**Comment by compnerd** [comment]
Sure, I can kick off a build for this

**Comment by compnerd** [comment]
For `_CFGetPathFromFileDescriptor`, something along the lines of the following should do the trick:
```c
HANDLE hFile = _get_osfhandle(fd);
WCHAR wszPath[MAX_PATH + 1];
DWORD dwLength = GetFinalPathNameByHandleW(hFile, &wszPath, ARRAY_SIZE(wszPath), FILE_NAME_NORMALIZED);
if (dwLength == 0)
  return false;
/* do `wcstombcs` dance */
```

**Comment by millenomi** [comment]
I'll let the first run go through so we can do a single pass with other Windows fixes as needed, but definitely will have a follow-up with that.

File: CoreFoundation/Base.subproj/CFRuntime.c [hunk 1]
```diff
@@ -1351,6 +1387,8 @@ int DllMain( HINSTANCE hInstance, DWORD dwReason, LPVOID pReserved ) {
 	// do these last
 	if (cfBundle) CFRelease(cfBundle);
         __CFStringCleanup();
+    } else if (dwReason == DLL_THREAD_DETACH) {
+        __CFFinalizeWindowsThreadData();
```
Comment (lxbndr) [inline CoreFoundation/Base.subproj/CFRuntime.c:1391]: This code was removed in #2902, as we are using modern thread local storage mechanics in Windows, which doesn't need explicit finalization.

File: CoreFoundation/NumberDate.subproj/CFDate.c [hunk 2]
```diff
@@ -171,10 +171,11 @@ CF_PRIVATE void __CFDateInitialize(void) {
...
 #elif TARGET_OS_WIN32
-    // We are using QueryUnbiasedInterruptTimePrecise as time source.
-    // It returns result in system time units of 100 nanoseconds. 
-    // To get seconds we need to divide the value by 1e7 (10000000).
-    __CFTSRRate = 1.0e7;
+    LARGE_INTEGER freq;
+    if (!QueryPerformanceFrequency(&freq)) {
+        HALT;
+    }
+    __CFTSRRate = (double)freq.QuadPart;
```
Comment (lxbndr) [inline CoreFoundation/NumberDate.subproj/CFDate.c:178]: This was modified in #2907. Long story short, CFRunLoop is sensitive to the way of using system timers. We have to keep it consistent with Dispatch, or CFRunLoop would rarely and randomly hang.

File: CoreFoundation/Base.subproj/CFRuntime.c (hunk 1 above)
Comment (millenomi) [reply CoreFoundation/Base.subproj/CFRuntime.c:1391]: Yup. The Windows build isn't erroring out at the undefined function; will remove.

File: CoreFoundation/NumberDate.subproj/CFDate.c (hunk 2 above)
Comment (millenomi) [reply CoreFoundation/NumberDate.subproj/CFDate.c:178]: Understood; will revert.

**Comment by millenomi** [comment]
@compnerd This is ready for another try on Windows.

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
cc @drexin This adds dependencies on new ICU symbols, do you see anything that may be alarming?

**Comment by YOCKOW** [comment]
@millenomi 
Sorry for a little digressing.
[You said, just one year ago](https://github.com/apple/swift-corelibs-foundation/pull/2870#issuecomment-688433305), that small CF changes were *temporarily* stopped from being merged. However, they are not merged yet as of now.
When should (or shouldn't) such changes be merged?

**Comment by millenomi** [comment]
It's been a hard year. The hope was to do this merge earlier, but we ended up skipping a release.

Once this patch is merged, by and large.

File: CoreFoundation/PlugIn.subproj/CFBundle_Internal.h [hunk 3]
```diff
@@ -378,6 +423,47 @@ CF_PRIVATE CFStringRef _CFBundleGetPlatformNameSuffix(void);
...
+#pragma mark -
+#pragma mark Resolving paths from FDs
+
+// The buffer must be PATH_MAX long or more.
+static bool _CFGetPathFromFileDescriptor(int fd, char *path);
+
+#if TARGET_OS_MAC || TARGET_OS_BSD
+
+static bool _CFGetPathFromFileDescriptor(int fd, char *path) {
+    return fcntl(fd, F_GETPATH, path) != -1;
```
[the remark this answers is not in the document]
Comment (millenomi) [reply CoreFoundation/PlugIn.subproj/CFBundle_Internal.h:435]: We already have a generic fallback; I'll just remove the `TARGET_OS_BSD`.

[the remark this answers is not in the document]
Comment (millenomi) [reply CoreFoundation/PlugIn.subproj/CFBundle_Internal.h:435]: … or condition this to `! __OpenBSD__`.

**Comment by millenomi** [comment]
@3405691582 I applied your suggestions, can you give another quick look?

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@3405691582 I think things are okay, and I'm happy to make follow-ups for BSD if needed, but I think my next priority is to close this up for Linux & Windows for now.

**Comment by 3405691582** [comment]
@millenomi I merged back your changes and it built OK so all is well anyway :) Thanks for applying the fixes!

There are some unrelated buildfixes I'll chase in other PRs.

**Comment by millenomi** [comment]
@3405691582 please ping me on those PRs and I'll expedite them.

**Comment by 3405691582** [comment]
Ah, it looks like my concern was dealt with by #3060 already, so I think we're good.

File: CoreFoundation/Base.subproj/CFAvailability.h [hunk 4]
```diff
@@ -135,13 +135,13 @@
...
-#define __CF_NAMED_ENUM(_type, _name)     int __CF_ENUM_ ## _name; enum __CF_ENUM_ATTRIBUTES _name : _type; typedef enum _name _name; enum _name : _type
+#define __CF_NAMED_ENUM(_type, _name)     enum __CF_ENUM_ATTRIBUTES _name : _type _name; enum _name : _type
 #define __CF_ANON_ENUM(_type)             enum __CF_ENUM_ATTRIBUTES : _type
-#define CF_CLOSED_ENUM(_type, _name)      int __CF_ENUM_ ## _name; enum __CF_CLOSED_ENUM_ATTRIBUTES _name : _type; typedef enum _name _name; enum _name : _type
+#define CF_CLOSED_ENUM(_type, _name)      enum __CF_CLOSED_ENUM_ATTRIBUTES _name : _type _name; enum _name : _type
 #if (__cplusplus)
 #define CF_OPTIONS(_type, _name) _type _name; enum __CF_OPTIONS_ATTRIBUTES : _type
 #else
-#define CF_OPTIONS(_type, _name) int __CF_OPTIONS_ ## _name; enum __CF_OPTIONS_ATTRIBUTES _name : _type; typedef enum _name _name; enum _name : _type
+#define CF_OPTIONS(_type, _name) enum __CF_OPTIONS_ATTRIBUTES _name : _type _name; enum _name : _type
```
Comment (buttaface) [inline CoreFoundation/Base.subproj/CFAvailability.h:144]: I had to revert these macro patches when building for Android, buttaface/swift-android-sdk@dba89d2, or [building CoreFoundation fails with this error](https://github.com/buttaface/swift-android-sdk/runs/3608214051?check_suite_focus=true#step:7:2268):
```
error: non-defining declaration of enumeration with a fixed underlying type is only permitted as a standalone declaration; missing list of enumerators? [-Welaborated-enum-base]
```
I checked the preprocessed source produced with and without these macro changes and got the following diff:
```
--- a/cf.i	2021-09-16 07:54:13.941791291 +0000
+++ b/cf.i	2021-09-16 08:08:32.447061538 +0000
@@ -199,7 +199,7 @@
 typedef __attribute__((objc_bridge(id))) CFTypeRef CFPropertyListRef;
 
 
-typedef int __CF_ENUM_CFComparisonResult; enum __attribute__((enum_extensibility(open))) CFComparisonResult : CFIndex; typedef enum CFComparisonResult CFComparisonResult; enum CFComparisonResult : CFIndex {
+typedef enum __attribute__((enum_extensibility(open))) CFComparisonResult : CFIndex CFComparisonResult; enum CFComparisonResult : CFIndex {
     kCFCompareLessThan = -1L,
     kCFCompareEqualTo = 0,
     kCFCompareGreaterThan = 1
```
I don't know what to make of that C gibberish so if anybody else knows how to fix it, I'd appreciate it.

**Pull request merged**